### DISCUSSION TITLE
JS `String` to `std::float` conversion

### DISCUSSION THREAD

**Discussion by jmarcosfer** [emscripten-core/emscripten#21479] [Q&A]
I have a class method in my Embind bindings that takes a `std::float` as an argument. However, the method doesn't throw when I pass it a `String` from JS as long as that `String` contains numeric characters. Is there some sort of implicit type casting done under the hood that I'm missing?

I've checked [this](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#built-in-type-conversions) part of the docs about type conversions, but the table makes it look like it's about C++ to JS conversions and not the other way around (plus with the behaviour I've found I know definitely that JS `String`s don't always become `std::string` on the other side).

For what it's worth, I've also found that JS `String`s with text characters are converted to `NaN`s on C++. 

Anything I'm missing? 
Thanks!

**Comment by sbc100** [comment]
I suspect this is just an artifact of the automatic conversion rules that apply when calling from JavaScript into Wasm. 

I believe the same thing would happen if you directly called an export C function from JS (i.e. without embind at all).

Assuming I'm correct, then the next question might be: Should emscripten attempt to do more strict type checking at the boundary than the normal JavaScript API does?  I guess it could provide more debugging in debug build perhaps?  @brendandahl @kripken WDYT?

**Reply by brendandahl** [reply]
We currently have a check for this if assertions are enabled https://github.com/emscripten-core/emscripten/blob/a589a22edd28ff6a3f1d6ac837cc55266c9379f3/src/embind/embind.js#L482-L486

**Reply by sbc100** [reply]
Oh nice.   @jmarcosfer can you confirm this is firing in a debug build for you?

@brendandahl @kripken should we perhaps move this check to be in a more generic place so it works the same for all Wasm exports (not just embind'd ones).

**Reply by jmarcosfer** [reply]
Thank you all for the quick reply.

The test I did used our "sync" build which has default `ASSERTIONS=1`. 
Should I try with `ASSERTIONS=2`?

Just for context, these are our `emcc` commands for building an "async" and a "sync" version.
```Makefile
@emcc -Oz $(BINDING_ESSENTIAJS) $(INCLUDE_ESSENTIAJS) \
	-lembind -lessentia -L $(LIB_DIR_ESSENTIA) \
	-I $(EIGEN_PATH) -I $(INCLUDE_DIR_ESSENTIA)\
	-s WASM=1 \
	-o $(ESSENTIA_JS_WEB) \
	-s EXCEPTION_DEBUG \
	-s ASSERTIONS=2 \
	-s ENVIRONMENT=web \
	-s MODULARIZE=1 \
	-s EXPORT_NAME="EssentiaWASM" \
	--post-js $(POST_JS_WEB_WASM) \
	-s ALLOW_MEMORY_GROWTH=1 || exit 1
@echo "Done ..."

@echo "... compiling sync builds..."
@emcc -Oz $(BINDING_ESSENTIAJS) $(INCLUDE_ESSENTIAJS) \
	-lembind -lessentia -L $(LIB_DIR_ESSENTIA) \
	-I $(EIGEN_PATH) -I $(INCLUDE_DIR_ESSENTIA)\
	-s WASM=1 \
	-o $(ESSENTIA_JS_MODULE) \
	-s BINARYEN_ASYNC_COMPILATION=0 \
	-s ALLOW_MEMORY_GROWTH=1 \
	-s SINGLE_FILE=1 || exit 1
```

PS: I realise this is probably suboptimal, so also feel free to suggest improvements.

**Reply by brendandahl** [reply]
The example below seems to work fine for me. Do you see the assertion code in the output JS file for your project?
```c++
#include <stdio.h>
#include <emscripten.h>
#include <emscripten/bind.h>

using namespace emscripten;

void foo(float f) {}

int main() {
EM_ASM(
  Module.foo("hello");
);
}

EMSCRIPTEN_BINDINGS(xxx) {
  function("foo", &foo);
}
```
built with:
```
emcc -Oz -sASSERTIONS=1 -lembind main.cpp -o a.out.js
```
gives me:
`TypeError: Cannot convert hello to undefined`
